src/history.rs

```diff
@@ -155,126 +153,184 @@ impl<'a> DynamicHistory<'a> {
+        }
+    }
+
+    pub(crate) fn report(&mut self, destination: &mut [u8]) -> Result<usize, ReportError> {
+        // If we can't store at least a header and one event, it's a hard error
+        if destination.len() < WireReport::<&[u8]>::buffer_len(0, 1) {
+            return Err(ReportError::InsufficientDestinationSize);
+        }
+
+        let self_clock = self.self_clock;
+        let mut report = WireReport::new_unchecked(destination);
+        report.set_fingerprint();
+        report.set_probe_id(self.probe_id);
+        report.set_clock(crate::pack_clock_word(self_clock.epoch, self_clock.ticks));
+
+        // We can't store at least the frontier clocks and a single two-word item
+        if report.as_ref().len() < WireReport::<&[u8]>::buffer_len(self.clocks.len(), 2) {
+            report.set_seq_num(self.report_seq_num);
+            report.set_n_clocks(0);
+            report.set_n_log_entries(1);
+            let payload = report.payload_mut();
+            payload[..size_of::<LogEntry>()].copy_from_slice(
+                &EventId::EVENT_INSUFFICIENT_REPORT_BUFFER_SIZE
+                    .get_raw()
+                    .to_le_bytes(),
+            );
+        } else {
+            let clocks_len = self.clocks.len();
+            report.set_seq_num(self.report_seq_num);
+            report.set_n_clocks(clocks_len as u16);
+
+            let mut payload = report.payload_mut();
+            let (n_items_written, n_items_read, did_clocks_overflow) = write_report_payload(
+                self.log.iter(self.read_cursor),
+                &mut FrontierClocksFSV::new(&mut self.clocks),
+                &mut PayloadOutputByteSlice::new(&mut payload),
+            );
+            // Buffer cannot store more than u32::MAX/2 items, so n_items_written will always fit into u32
+            report.set_n_log_entries(n_items_written as u32);
+
+            // Wrapping add
+            // Buffer cannot store more than u32::MAX/2 items, so n_items_read will always fit into u32
+            self.read_cursor = self.log.seqn_add(self.read_cursor, n_items_read as u32);
+            if did_clocks_overflow {
+                self.record_event(EventId::EVENT_NUM_CLOCKS_OVERFLOWED);
+            }
         }
-        self.increment_local_clock_count();
-        self.write_current_clocks_to_log();
-        Ok(CausalSnapshot {
-            clock: self.clocks[0],
+
+        self.report_seq_num = self.report_seq_num.wrapping_add(1);
+        self.record_event(EventId::EVENT_PRODUCED_EXTERNAL_REPORT);
+
+        Ok(WireReport::<&[u8]>::header_len() + report.payload_len())
+    }
+
+    #[inline]
+    pub(crate) fn produce_snapshot(&mut self) -> Result<CausalSnapshot, ProduceError> {
+        let snap = CausalSnapshot {
+            clock: self.self_clock,
             reserved_0: 0,
             reserved_1: 0,
-        })
+        };
+        self.increment_local_clock();
+        self.write_clocks_to_log(&[self.self_clock]);
```

Comment:
I think now that we only ever write one clock to the log at a time, I think we ought to transform this fn to `write_clock_to_log`.